File: backend/app/schemas/academic.py
"""Schemas for Faculty Academic Workflow (Assignments, Attendance, Student Roster)."""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

# --- Assignment Schemas ---


class AssignmentCreateIn(BaseModel):
    course_code: str = Field(..., min_length=1, max_length=50)
    course_name: str = Field(..., min_length=1, max_length=255)
    title: str = Field(..., min_length=1, max_length=255)
    description: str = Field(..., min_length=1)
    due_date: datetime
    total_points: int = Field(default=100, ge=1, le=1000)


class AssignmentSubmissionOut(BaseModel):
    id: uuid.UUID
    assignment_id: uuid.UUID
    student_id: uuid.UUID
    student_email: str
    submission_text: str
    score: int | None
    feedback: str | None
    status: str
    submitted_at: datetime
    graded_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class AssignmentOut(BaseModel):
    id: uuid.UUID
    faculty_id: uuid.UUID
    course_code: str
    course_name: str
    title: str
    description: str
    due_date: datetime
    total_points: int
    status: str
    created_at: datetime
    submission_count: int = 0
    graded_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class AssignmentDetailOut(AssignmentOut):
    submissions: list[AssignmentSubmissionOut] = []


class SubmissionGradeIn(BaseModel):
    submission_id: uuid.UUID
    score: int = Field(..., ge=0, le=1000)
    feedback: str | None = None


# --- Attendance Schemas ---


class AttendanceRecordIn(BaseModel):
    student_id: uuid.UUID
    student_email: str
    status: str = Field(default="present", pattern="^(present|absent|late|excused)$")
    notes: str | None = None


class AttendanceSessionCreateIn(BaseModel):
    course_code: str = Field(..., min_length=1, max_length=50)
    course_name: str = Field(..., min_length=1, max_length=255)
    session_date: date
    topic: str = Field(..., min_length=1, max_length=255)
    records: list[AttendanceRecordIn]


class AttendanceRecordOut(BaseModel):
    id: uuid.UUID
    session_id: uuid.UUID
    student_id: uuid.UUID
    student_email: str
    status: str
    notes: str | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AttendanceSessionOut(BaseModel):
    id: uuid.UUID
    faculty_id: uuid.UUID
    course_code: str
    course_name: str
    session_date: date
    topic: str
    created_at: datetime
    present_count: int = 0
    absent_count: int = 0
    late_count: int = 0
    total_students: int = 0
    records: list[AttendanceRecordOut] = []

    model_config = ConfigDict(from_attributes=True)


# --- Student Roster Directory Schemas ---


class StudentRosterItemOut(BaseModel):
    id: uuid.UUID
    email: str
    full_name: str
    student_id: str
    program: str
    semester: str
    attendance_percentage: float
    total_sessions_attended: int
    total_sessions: int
    assignments_submitted: int
    total_assignments: int

    model_config = ConfigDict(from_attributes=True)


# --- Student Marks & Enrollment Lookup Schemas ---


class StudentMarkOut(BaseModel):
    id: uuid.UUID
    subject_code: str
    subject_name: str
    semester: str
    internal_marks: float
    midterm_marks: float
    final_marks: float
    total_marks: float
    grade: str
    grade_points: int
    credits: int
    academic_year: str

    model_config = ConfigDict(from_attributes=True)


class StudentAssignmentSubmissionItem(BaseModel):
    assignment_id: uuid.UUID
    course_code: str
    course_name: str
    title: str
    total_points: int
    due_date: datetime
    submission_id: uuid.UUID | None = None
    submission_text: str | None = None
    score: int | None = None
    feedback: str | None = None
    status: str = "pending"
    submitted_at: datetime | None = None


class StudentAcademicLookupOut(BaseModel):
    student_id: uuid.UUID
    enrollment_number: str
    name: str
    email: str
    phone_number: str | None = None
    branch: str | None = None
    course: str | None = None
    semester: str | None = None
    marks: list[StudentMarkOut] = []
    spi: float | None = None
    cpi: float | None = None
    assignments: list[StudentAssignmentSubmissionItem] = []
    attendance_percentage: float = 92.5
    total_sessions_attended: int = 0
    total_sessions: int = 0


# --- Student Exam Form Schemas ---


class ExamFormStatusOut(BaseModel):
    is_active: bool
    session_name: str
    announcement: str | None = None
    fee_amount: int = 1200
    start_date: datetime | None = None
    end_date: datetime | None = None


class ExamFormRegisterIn(BaseModel):
    semester: str = "Semester 6"
    papers: list[dict[str, str]]
    payment_reference: str | None = None

File: frontend/lib/api/academic.ts
/**
 * Academic API Client: Assignments, Attendance, and Student Roster.
 */

import { apiFetch } from "@/lib/api/client";

export interface AssignmentSubmissionOut {
  id: string;
  assignment_id: string;
  student_id: string;
  student_email: string;
  submission_text: string;
  score: number | null;
  feedback: string | null;
  status: "submitted" | "graded";
  submitted_at: string;
  graded_at: string | null;
}

export interface AssignmentOut {
  id: string;
  faculty_id: string;
  course_code: string;
  course_name: string;
  title: string;
  description: string;
  due_date: string;
  total_points: number;
  status: string;
  created_at: string;
  submission_count: number;
  graded_count: number;
}

export interface AssignmentDetailOut extends AssignmentOut {
  submissions: AssignmentSubmissionOut[];
}

export interface AssignmentCreateIn {
  course_code: string;
  course_name: string;
  title: string;
  description: string;
  due_date: string;
  total_points: number;
}

export interface SubmissionGradeIn {
  submission_id: string;
  score: number;
  feedback?: string;
}

export interface AttendanceRecordIn {
  student_id: string;
  student_email: string;
  status: "present" | "absent" | "late" | "excused";
  notes?: string;
}

export interface AttendanceRecordOut {
  id: string;
  session_id: string;
  student_id: string;
  student_email: string;
  status: "present" | "absent" | "late" | "excused";
  notes: string | null;
  created_at: string;
}

export interface AttendanceSessionOut {
  id: string;
  faculty_id: string;
  course_code: string;
  course_name: string;
  session_date: string;
  topic: string;
  created_at: string;
  present_count: number;
  absent_count: number;
  late_count: number;
  total_students: number;
  records: AttendanceRecordOut[];
}

export interface AttendanceSessionCreateIn {
  course_code: string;
  course_name: string;
  session_date: string;
  topic: string;
  records: AttendanceRecordIn[];
}

export interface StudentRosterItemOut {
  id: string;
  email: string;
  full_name: string;
  student_id: string;
  program: string;
  semester: string;
  attendance_percentage: number;
  total_sessions_attended: number;
  total_sessions: number;
  assignments_submitted: number;
  total_assignments: number;
}

// API Functions

export function listAssignments(accessToken: string): Promise<AssignmentOut[]> {
  return apiFetch<AssignmentOut[]>("/faculty/academic/assignments", { accessToken });
}

export function createAssignment(
  accessToken: string,
  payload: AssignmentCreateIn
): Promise<AssignmentOut> {
  return apiFetch<AssignmentOut>("/faculty/academic/assignments", {
    method: "POST",
    body: payload,
    accessToken,
  });
}

export function getAssignmentDetail(
  accessToken: string,
  assignmentId: string
): Promise<AssignmentDetailOut> {
  return apiFetch<AssignmentDetailOut>(`/faculty/academic/assignments/${assignmentId}`, {
    accessToken,
  });
}

export function deleteAssignment(accessToken: string, assignmentId: string): Promise<void> {
  return apiFetch<void>(`/faculty/academic/assignments/${assignmentId}`, {
    method: "DELETE",
    accessToken,
  });
}

export function gradeSubmission(
  accessToken: string,
  assignmentId: string,
  payload: SubmissionGradeIn
): Promise<AssignmentSubmissionOut> {
  return apiFetch<AssignmentSubmissionOut>(`/faculty/academic/assignments/${assignmentId}/grade`, {
    method: "POST",
    body: payload,
    accessToken,
  });
}

export function listAttendanceSessions(accessToken: string): Promise<AttendanceSessionOut[]> {
  return apiFetch<AttendanceSessionOut[]>("/faculty/academic/attendance/sessions", {
    accessToken,
  });
}

export function recordAttendanceSession(
  accessToken: string,
  payload: AttendanceSessionCreateIn
): Promise<AttendanceSessionOut> {
  return apiFetch<AttendanceSessionOut>("/faculty/academic/attendance/sessions", {
    method: "POST",
    body: payload,
    accessToken,
  });
}

export function getStudentRoster(accessToken: string): Promise<StudentRosterItemOut[]> {
  return apiFetch<StudentRosterItemOut[]>("/faculty/academic/students", { accessToken });
}

// ---------------------------------------------------------------------------
// Student Academic Records & Enrollment Lookup (Faculty & Admin)
// ---------------------------------------------------------------------------

export interface StudentMarkOut {
  id: string;
  subject_code: string;
  subject_name: string;
  semester: string;
  internal_marks: number;
  midterm_marks: number;
  final_marks: number;
  total_marks: number;
  grade: string;
  grade_points: number;
  credits: number;
  academic_year: string;
}

export interface StudentAssignmentSubmissionItem {
  assignment_id: string;
  course_code: string;
  course_name: string;
  title: string;
  total_points: number;
  due_date: string;
  submission_id?: string | null;
  submission_text?: string | null;
  score?: number | null;
  feedback?: string | null;
  status: string;
  submitted_at?: string | null;
}

export interface StudentAcademicLookupOut {
  student_id: string;
  enrollment_number: string;
  name: string;
  email: string;
  phone_number?: string | null;
  branch?: string | null;
  course?: string | null;
  semester?: string | null;
  marks: StudentMarkOut[];
  spi?: number | null;
  cpi?: number | null;
  assignments: StudentAssignmentSubmissionItem[];
  attendance_percentage: number;
  total_sessions_attended: number;
  total_sessions: number;
}

export function getStudentByEnrollment(
  accessToken: string,
  enrollmentNumber: string
): Promise<StudentAcademicLookupOut> {
  return apiFetch<StudentAcademicLookupOut>(
    `/faculty/academic/students/by-enrollment/${encodeURIComponent(enrollmentNumber)}`,
    { accessToken }
  );
}

// ---------------------------------------------------------------------------
// Examination Form (Student & Public Status)
// ---------------------------------------------------------------------------

export interface ExamFormStatusOut {
  is_active: boolean;
  session_name: string;
  announcement?: string | null;
  fee_amount: number;
  start_date?: string | null;
  end_date?: string | null;
}

export interface ExamFormRegisterIn {
  semester: string;
  papers: Array<{ code: string; title: string; type?: string }>;
  payment_reference?: string;
}

export function getExamFormStatus(accessToken: string): Promise<ExamFormStatusOut> {
  return apiFetch<ExamFormStatusOut>("/academic/exam-form/status", {
    accessToken,
  });
}

export function registerExamForm(
  accessToken: string,
  payload: ExamFormRegisterIn
): Promise<{ success: boolean; registration_id: string; message: string }> {
  return apiFetch<{ success: boolean; registration_id: string; message: string }>(
    "/academic/exam-form/register",
    {
      method: "POST",
      body: payload,
      accessToken,
    }
  );
}

